python/kvikio/numpy.py

```diff
@@ -0,0 +1,188 @@
+
+    >>> cupy.fromfile("/tmp/myfile", dtype=int, like=cupy.empty(()))
+    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+    >>> type(_)
+    <class 'cupy._core.core.ndarray'>
+    """
+
+    def __init__(self, like: ArrayLike) -> None:
+        self._like = like
+
+    def __array_function__(self, func, types, args, kwargs):
+        if func is not np.fromfile:
+            return func(*args, **kwargs)
+        return fromfile(*args, like=self._like, **kwargs)
+
+
+def fromfile(
+    file: Union[str, os.PathLike, io.FileIO],
+    dtype: DTypeLike = float,
+    count: int = -1,
+    sep: str = "",
+    offset: int = 0,
+    *,
+    like: ArrayLike = None,
+) -> ArrayLike:
+    """Construct an array from file using KvikIO
+
+    Overload `numpy.fromfile` to use KvikIO.
+
+    Parameters
+    ----------
+    file : FileLike or str or PathLike
+        Open file object or filename.
+    dtype : data-type
+        Data type of the returned array.
+        For binary files, it is used to determine the size and byte-order
+        of the items in the file.
+        Most builtin numeric types are supported and extension types may be supported.
+    count : int
+        Number of items to read. `-1` means all items (i.e., the complete file).
+    sep : str
+        Empty ("") separator means the file should be treated as binary. Any other
+        value is not supported and will raise NotImplementedError.
+    offset : int
+        The offset (in bytes) from the file's current position. Defaults to 0.
+        Only permitted for binary files.
+    like : array_like, optional
+        Reference object to allow the creation of arrays which are not
+        NumPy arrays.
+
+    Examples
+    --------
+    Read file into a NumPy array:
+
+    >>> np.arange(10).tofile("/tmp/myfile")
+    >>> fromfile("/tmp/myfile", dtype=int)
+    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+    >>> type(_)
+    <class 'numpy.ndarray'>
+
+    Read file into a CuPy array using the like argument. The file is read
+    directly into device memory using GDS if available:
+
+    >>> import cupy
+    >>> fromfile("/tmp/myfile", dtype=int, like=cupy.empty(()))
+    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+    >>> type(_)
+    <class 'cupy._core.core.ndarray'>
+    """
+
+    if sep != "":
+        raise NotImplementedError(
+            "Non-default value of the `sep` argument is not supported"
+        )
+
+    if isinstance(file, FileLike):
+        filepath = file.name
+    else:
+        filepath = str(file)
+
+    nbytes = os.path.getsize(filepath)
+    itemsize = np.dtype(dtype).itemsize
+    if count == -1:
+        count = nbytes // itemsize
+        if nbytes % itemsize != 0:
+            raise ValueError(f"file size ({nbytes}) not divisible with dtype ({dtype})")
+    if count * itemsize > nbytes:
+        raise ValueError(f"count ({count*itemsize}) greater than file size ({nbytes})")
+
+    # Truncate to itemsize divisible
+    count -= math.ceil(offset / itemsize)
```

Comment:
In playing with `fromfile` and some fits data I'm ran into an issue with `count`:

```python
# wget https://fits.gsfc.nasa.gov/samples/UITfuv2582gc.fits
offset = 11520
count = 512**2
shape = (512, 512)
data = cp.fromfile('UITfuv2582gc.fits', dtype=np.dtype('int16'), offset=offset, count=count, like=LikeWrapper(cp.empty(()))
len(data)
```
`data` is length `256384` and should be 512**2 (262144) .  Do recall why truncation is needed ?

Reply:
Isn't there a header in [the FITS file format]( https://fits.gsfc.nasa.gov/fits_primer.html )?

Also curious why `int16` was chosen here as opposed to something like `uint8` followed by splitting up the data (based on the format) and then casting as appropriate?

Reply:
There is a header, that's why I'm also including `offset` .  int16 was chosen because I'm more or less cribbing this from the astropy.fits work and this is what they are using to read in the file

Reply:
Should be fixed in https://github.com/rapidsai/kvikio/pull/135/commits/7fb27a169c7ebc9ae1ebf319b796b774a703264d

Reply:
@quasiben does it work for you?

Reply:
Yes!  Thank you for the quick fix!